lib/c.c

```diff
@@ -370,12 +370,89 @@ int fputc(int c, FILE *stream)
  * FIXME: adopt lite_malloc from musl-libc
  *   https://git.musl-libc.org/cgit/musl/tree/src/malloc/lite_malloc.c
  */
-void *malloc(int size)
+
+typedef struct block_meta {
+    int size;
+    struct block_meta *next;
+    int free;
+} block_meta_t;
+
+block_meta_t *__malloc_global_base;
+block_meta_t *__malloc_global_last;
+
+block_meta_t *get_block_ptr(void *ptr)
 {
-    char *brk = __syscall(__syscall_brk, 0); /* read current break */
-    char *request = __syscall(__syscall_brk, brk + size);
+    return ptr - sizeof(block_meta_t);
+}
+
+block_meta_t *find_free_block(int size)
+{
+    block_meta_t *current = __malloc_global_base;
+    while (current != NULL) {
+        /* TODO: support break in while loop */
+        if (current->free == 1 && current->size >= size)
+            return current;
+        __malloc_global_last = current;
+        current = current->next;
+    }
+    return current;
+}
+
+block_meta_t *request_space(int size)
```

Comment:
Function `request_space` would be called multiple times. Let's add pseudo namespace. That is, specify prefix `__malloc_` for this function to avoid potential symbol conflicts.